### PR TITLE
fallback to proxy config global protocol when upstream services' protocol is unset

### LINKED ISSUE #5857: connect: allow setting a default protocol for all services in `proxy-defaults`
It would be convenient for some users who mostly deploy HTTP services to make this the default for services rather than having to manually opt in to something other than `tcp`.

This is _almost_ possible already by setting the protocol in the global `proxy-defaults`:

```hcl
kind = "proxy-defaults"
name = "global"
config {
  protocol = "http"
}
```

Doing this would get automatically resolved by all service and so they would all get that protocol set in their proxy config and so get an HTTP-aware public listener installed in the proxy.

But currently in `ResolveServiceConfig` RPC, when we are looking up the upstream's configs, we only look in their `service-defaults` for a protocol and if it's not set default to nothing (implying TCP).

We just need to:
 - In `ResolveServiceConfig` validate that global protocol is picked up for `ProxyConfig['protocol']` when `service-defaults` does not set one
 - In `ResolveServiceConfig` if an upstream has no config or no protocol set in it, use the global-defaults protocol value instead of empty.

This should be a really simple change.

### PR DESCRIPTION
**Pull request by adams-sarah**
Fixes #5857

### CODE REVIEW

File: agent/consul/config_endpoint_test.go [hunk 1]
```diff
@@ -704,6 +704,65 @@ func TestConfigEntry_ResolveServiceConfig(t *testing.T) {
...
+		Name: structs.ProxyConfigGlobal,
+		Config: map[string]interface{}{
+			"protocol": "http",
+		},
+	}))
+	require.NoError(state.EnsureConfigEntry(2, &structs.ServiceConfigEntry{
+		Kind: structs.ServiceDefaults,
+		Name: "foo",
+	}))
+	require.NoError(state.EnsureConfigEntry(2, &structs.ServiceConfigEntry{
```
Comment (freddygv) [inline agent/consul/config_endpoint_test.go]: I think this test could benefit from adding an upstream config entry that does have a protocol set. That way we can also ensure that it takes precedence over the global default.

File: agent/consul/config_endpoint.go [hunk 2]
```diff
@@ -289,16 +295,27 @@ func (c *ConfigEntry) ResolveServiceConfig(args *structs.ServiceConfigRequest, r
...
 				}
 
+				// No upstream found; skip.
+				if upstreamConf == nil {
+					continue
+				}
+
+				// Fallback to proxyConf global protocol.
+				protocol := proxyConfGlobalProtocol
+				if upstreamConf != nil && upstreamConf.Protocol != "" {
```
[the remark this answers is not in the document]
Comment (adams-sarah) [reply agent/consul/config_endpoint.go]: aah shoot!! awesome catch thank you @freddygv ! will do

File: agent/consul/config_endpoint_test.go (hunk 1 above)
Comment (adams-sarah) [reply agent/consul/config_endpoint_test.go]: doii great idea thank you!

**Comment by adams-sarah** [comment]
Thanks for the review @freddygv ! altered PR per your suggestions. PTAL!

**Pull request merged**